### PR TITLE
help wanted: review Z80 interruptible 

### PR DESCRIPTION
**Pull request by holub**
Basic mechanics is implemented.

Details:
Instruction defined as ```ops_type = vector<function<void>>```
Instruction execution controlled by state stored in _m_prefix_, _m_opcode_, and _m_cycle_.

Next:
* **confirm approach is acceptable**
* split instructions more granular - at r/w first, then at any calls with side-effect

Problems:
Performance - ~~even in current state (with almost no steps spitted) we lost 40-50%. Expecting worse with more granular.~~
UPDATE: Performance issues are noticeable in debug. For real build it's not as significant.

Ideas to discuss:
* hold instruction definition in array instead var
* use array instead of vector for instruction definition
* make sure instruction steps are return by reference on access, not copied from definition
* ???

### CODE REVIEW

**Comment by rb6502** [comment]
The 6502 family cores achieve cycle-interruptible operation without a major performance hit.  @galibert can explain how he did it.

**Comment by holub** [comment]
> The 6502 family cores achieve cycle-interruptible operation without a major performance hit. @galibert can explain how he did it.

Thanks rb, we already talked and I'm familiar with 6502 implementation.
The goal hear was to make changes with less distraction to the current code base.
And honestly I have concerns that 6502 approach will behave much faster than this PR (assuming undesired mem allocations fully removed) compare to original state where all commands executed in single shot and most of the code inlined.

File: src/devices/cpu/z80/z80.cpp [hunk 1]
```diff
@@ -1431,1767 +1439,1776 @@ inline void z80_device::ei()
...
+OP(cb,fc) { H = set(7, H);          } EOP /* SET  7,H         */
+OP(cb,fd) { L = set(7, L);          } EOP /* SET  7,L         */
+OP(cb,fe) { wm(HL, set(7, rm_reg(HL))); } EOP /* SET  7,(HL)      */
+OP(cb,ff) { A = set(7, A);          } EOP /* SET  7,A         */
 
 
 /**********************************************************
 * opcodes with DD/FD CB prefix
-* rotate, shift and bit operations with (IX+o)
+* rotate, shift and bit EOPrations with (IX+o)
```
Comment (qeeg) [inline src/devices/cpu/z80/z80.cpp:1756]: Copy paste error

File: src/devices/cpu/z80/z80.cpp [hunk 2]
```diff
@@ -3473,6 +3492,7 @@ void z80_device::device_start()
 	save_item(NAME(m_busrq_state));
 	save_item(NAME(m_after_ei));
 	save_item(NAME(m_after_ldair));
+	save_item(NAME(m_m_shared));
```
Comment (qeeg) [inline src/devices/cpu/z80/z80.cpp:3495]: Is this supposed to have two `m_` prefixes on it?

Comment (holub) [reply src/devices/cpu/z80/z80.cpp:3495]: Yes m-cycle of instruction - var shared between M-states

Comment (qeeg) [reply src/devices/cpu/z80/z80.cpp:3495]: Ah okay.

File: src/devices/cpu/z80/z80.cpp (hunk 1 above)
Comment (holub) [reply src/devices/cpu/z80/z80.cpp:1756]: thanks. fixed locally, will commit with next update

**Comment by holub** [comment]
Hmm. Just noticed that performance problem is with debug. Without debug - all blazing bast.

**Comment by holub** [comment]
So at this point I'm looking for approval of this approach.
If received, I'll work on further changes to make it completed.

**Pull request closed without merging**